### PR TITLE
[RFC] DBT freshness check builder

### PR DESCRIPTION
**Pull request by dpeng817**
Creates freshness checks from dbt config. The idea is that users of dbt can define their freshnes dependencies without ever needing to directly interface with python, and the platform owner can just call one function to construct all freshness checks easily. This will scale as more models get added, without needing to write additional python.

I went with this solution because it seemed like the shortest path towards the requirement of allowing data team members to encode freshness in dbt rather than python, with minimal python required. There are notable gaps with this solution; namely the lack of support for setting freshness checks across larger swaths of models. I think for more complex use cases like that, we defer back to the python layer, and improve the ergonomics for asset selection passing to a freshness check builder, for example.

Other potential solutions, and reasons why I didn't choose them for this initial build:
- Make the `@dbt_assets` decorator support returning more than just a single `Definitions` object. This would entail a breaking change, basically no way to do something like this with our current roadmap.
- Provide some way of generating additional definitions from an AssetsDefinition. Specifically, a lambda or something that can construct the freshness checks and sensor at `Definitions` construction. This seems really hacky and somewhat adjacent to the maligned CacheableAssetsDefinition. I have doubts that we could get something like this over the wire.
- Do nothing, and expect users to just use python. I think that this will overall block adoption, and should be avoided if possible.

How I tested this:
I added unit tests for the partitioned, non-partitioned, and mixed-partitioning case. I also added testing for providing multiple assets definitions which utilize the same manifest, and providing duplicate asset keys.

### CODE REVIEW

**Comment by dpeng817** [comment]
* **#22290** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22290?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @dpeng817 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/libraries/dagster-dbt/dagster_dbt_tests/conftest.py [hunk 1]
```diff
@@ -17,8 +17,10 @@
     test_dbt_unit_tests_path,
     test_duplicate_source_asset_key_path,
     test_jaffle_shop_path,
+    test_last_update_freshness_path,
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt_tests/conftest.py:20]: To make this easier to review, could you add the two base jaffle shop projects in a separate PR, similar to https://github.com/dagster-io/dagster/pull/22246?

It's hard to tell what has changed in the dbt projects that you're testing.

File: python_modules/libraries/dagster-dbt/dagster_dbt/freshness_builder.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+from .asset_utils import get_manifest_and_translator_from_dbt_assets
+
+
+def build_freshness_checks_from_dbt_assets(
+    dbt_assets: Sequence[AssetsDefinition],
+) -> Sequence[AssetChecksDefinition]:
+    """Returns a sequence of freshness checks constructed from the provided dbt assets.
+
+    Freshness checks can be configured on a per-model basis in the dbt project configuration.
+    For time-partitioned assets, the freshness check configuration mirrors that of the `build_time_partition_freshness_checks` function.
```
Comment (sryza) [inline python_modules/libraries/dagster-dbt/dagster_dbt/freshness_builder.py:30]: Nit: line length

Also, a newline after this line would make this easier to read

File: python_modules/libraries/dagster-dbt/dagster_dbt/freshness_builder.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+                )
+            else:
+                try:
+                    lower_bound_delta_seconds_str = freshness_check_config[
+                        "lower_bound_delta_seconds"
+                    ]
+                    lower_bound_seconds = int(lower_bound_delta_seconds_str)
+                except:
+                    raise DagsterInvariantViolationError(
+                        "lower_bound_delta_seconds must be provided, and parseable as an integer in the freshness_check config for a non-time-partitioned asset."
```
Comment (sryza) [inline python_modules/libraries/dagster-dbt/dagster_dbt/freshness_builder.py:103]: line length

File: python_modules/libraries/dagster-dbt/dagster_dbt/freshness_builder.py (hunk 2 above)
Comment (sryza) [inline python_modules/libraries/dagster-dbt/dagster_dbt/freshness_builder.py:30]: Non-partitioned dbt assets are quite a bit more common than partitioned dbt assets, so we should lead with the former

File: python_modules/libraries/dagster-dbt/dagster_dbt/freshness_builder.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    models:
+      - name: customers
+        ...
+        meta:
+          dagster:
+            freshness_check:
+              deadline_cron: "0 0 * * *"
+              severity: "WARN" # Optional, defaults to "WARN"
+    ```
+    For non-time-partitioned assets, the freshness check configuration mirrors that of the
```
Comment (sryza) [inline python_modules/libraries/dagster-dbt/dagster_dbt/freshness_builder.py:42]: Need some newlines between these sections

File: python_modules/libraries/dagster-dbt/dagster_dbt/freshness_builder.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    `build_last_update_freshness_checks` function. `lower_bound_delta` is provided in terms of
+    seconds, and `deadline_cron` is optional.
+    As an example:
+    ```yaml
```
Comment (sryza) [inline python_modules/libraries/dagster-dbt/dagster_dbt/freshness_builder.py:46]: We typically do docstring examples like this:
```
    Examples:
        .. code-block:: yaml
```

**Review by sryza — CHANGES_REQUESTED** [review]
@slopp - would you be up for taking a look at this one, as you were a strong advocate for it if I recall?

Also, we should make sure to expose this publicly

File: python_modules/libraries/dagster-dbt/dagster_dbt/freshness_builder.py (hunk 5 above)
Comment (dpeng817) [reply python_modules/libraries/dagster-dbt/dagster_dbt/freshness_builder.py:46]: Aware; but wasn't sure if this quite fit that mold. I guess I had some preconception that code examples of that format would fit the same language pattern as the API in question, but perhaps not.

Comment (sryza) [reply python_modules/libraries/dagster-dbt/dagster_dbt/freshness_builder.py:46]: > I guess I had some preconception that code examples of that format would fit the same language pattern as the API in question

I don't think that's a prerequisite

File: python_modules/libraries/dagster-dbt/dagster_dbt/freshness_builder.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+from dagster._core.errors import DagsterInvariantViolationError
+
+from dagster_dbt.utils import ASSET_RESOURCE_TYPES
+
+from .asset_utils import get_manifest_and_translator_from_dbt_assets
+
+
+def build_freshness_checks_from_dbt_assets(
+    dbt_assets: Sequence[AssetsDefinition],
+) -> Sequence[AssetChecksDefinition]:
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/freshness_builder.py:26]: I'd be wary of accepting a sequence of `dbt_assets` here.

If a user has `n` `dbt_assets`, we'd potentially define the same asset check definitions `n` times.

We should:
- enforce that `dbt_assets` is singular
- let this optionally take in a manifest and `dagster_dbt_translator` instead.

**Review by rexledesma — DISMISSED** [review]
I don't think the current API will work in a world where a user has two dbt assets definitions: one for partitioned assets, and one for non-partitioned assets. We'd create duplicate checks.

We should probably reverse index the manifest by asset key, and iterate on the asset keys defined on the passed definitions instead.

**Comment by sryza** [comment]
> We should probably reverse index the manifest by asset key, and iterate on the asset keys defined on the passed definitions instead.

oo great point

> let this optionally take in a manifest and dagster_dbt_translator instead.

Why exactly?

**Comment by rexledesma** [comment]
> > let this optionally take in a manifest and dagster_dbt_translator instead.
> 
> Why exactly?

See this interaction with a user:
image

The pattern of using dbt_assets in these utility functions doesn't scale when:

- You start using Dagster core utility functions like `load_assets_from_package_module`.
- You start creating multiple asset definitions from the same manifest (e.g. think of defining incremental models in a separate definition from non-incremental models)

We should begin to divest from passing in `@dbt_assets` to `get_asset_key_for_model`, ..., `build_schedule_from_dbt_selection`.
Instead, we should just make them pass in `manifest` and `dagster_dbt_translator` explicitly. Or at the very least, add them as optional arguments so they have an escape hatch if they encounter circular dependencies, etc.

File: python_modules/libraries/dagster-dbt/dagster_dbt/freshness_builder.py (hunk 6 above)
Comment (dpeng817) [reply python_modules/libraries/dagster-dbt/dagster_dbt/freshness_builder.py:26]: Do we have an easy way for users to retrieve the manifest and translator / are those objects readily available? I would like to avoid the situation where there are two sets of valid parameters, and all parameters are optional on the signature (high parameter anxiety), so if we feel like the manifest and translator are more scalable params, then let's just go with those. @sryza do you have any thoughts to the contrary here?

**Comment by sryza** [comment]
@rexledesma for that issue referenced in the image above, if I understand correctly it can come up even in non-dbt situations, which leads me to think we should try to find a more general solution for it.

I just reproduced it in the following way:

assets1.py
```python
from dagster import asset


@asset
def asset1(): ...
```

assets2.py
```python
from dagster import asset

from .assets1 import asset1


@asset(deps=[asset1])
def asset2(): ...
```

assets3.py
```python
from dagster import asset

from .assets1 import asset1


@asset(deps=[asset1])
def asset3(): ...
```

definitions.py
```python
from assets import assets2, assets3
from dagster import Definitions, load_assets_from_modules

assets2 = load_assets_from_modules([assets2])
assets3 = load_assets_from_modules([assets3])


defs = Definitions(assets=[*assets2, *assets3])
```

Produces:
```
dagster._core.errors.DagsterInvalidDefinitionError: Duplicate asset key: AssetKey(['asset1'])
```

I think the best solution here would probably be, in the `Definitions` constructor, to dedupe `AssetsDefinitions` that have reference equality.

File: python_modules/libraries/dagster-dbt/dagster_dbt/freshness_builder.py (hunk 6 above)
Comment (rexledesma) [reply python_modules/libraries/dagster-dbt/dagster_dbt/freshness_builder.py:26]: > I understand why we would be defining the same asset check definitions n times - is the concern that users will pass in the same AssetsDefinition object n times? Or am I missing something.

The current implementation reads the manifest from the `@dbt_assets` definition and constructs a freshness check for each `dbt_resource_props` in the manifest.

If you pass in two different `@dbt_assets` definitions that work off of the same manifest, then you would have created the same freshness checks twice.

Comment (dpeng817) [reply python_modules/libraries/dagster-dbt/dagster_dbt/freshness_builder.py:26]: right - I'm saying once that problem is solved / we're operating on the list of asset keys directly, we should be able to pass a sequence, or is there anything else?

Comment (rexledesma) [reply python_modules/libraries/dagster-dbt/dagster_dbt/freshness_builder.py:26]: Sure, let's still pass in a sequence of definitions so that this API is inline with the other signatures.

We should de-dup the parameters passed in `Definitions` as a follow up.

**Comment by dpeng817** [comment]
Hey @rexledesma and @sryza ;
I think I've addressed all the outstanding feedback. I've changed the API to explicitly check against the asset keys provided in the assets definition rather than rely on the manifest, and I've added tests for the selection against the same manifest behavior.

**Comment by slopp** [comment]
The gist of this looks ok to me, I am attempting to test in hooli: https://github.com/dagster-io/hooli-data-eng-pipelines/pull/96

Two main questions:
- Do I need to create a sensor to evaluate these checks?
- Do we plan to add docs? (It was a little hard to parse the tests to determine how to add all of this)

A set of eyes on the hooli PR would also be a good gut check

**Comment by dpeng817** [comment]
@slopp ; plan is to add docs in a follow-on PR.

**Review by rexledesma — APPROVED** [review]
Code looks fine to me. I would double-check that the docstrings properly render.

---

We discussed this offline, but another thing to explicitly note here is that by relying on the dbt config system, the only way to apply Dagster freshness checks to a collection of dbt models is in a hierarchical fashion, according to the dbt project's project structure.

If a user wanted to say, apply a Dagster freshness check to all models with a dbt tag named "strict", they would be unable to do that in this scheme.

It could be fine to ignore this potential use-case for now, as we just want initial adoption for this feature. But in the future, we should determine if this is a legitimate use case or not. If it is, how will our current API support it?
